servlet/src/main/java/io/grpc/servlet/ServerBuilder.java

```diff
@@ -0,0 +1,198 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.grpc.servlet;
+
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.base.Preconditions.checkState;
+
+import com.google.common.util.concurrent.ListenableFuture;
+import io.grpc.Server;
+import io.grpc.ServerStreamTracer.Factory;
+import io.grpc.Status;
+import io.grpc.internal.AbstractServerImplBuilder;
+import io.grpc.internal.Channelz.SocketStats;
+import io.grpc.internal.GrpcUtil;
+import io.grpc.internal.Instrumented;
+import io.grpc.internal.InternalServer;
+import io.grpc.internal.LogId;
+import io.grpc.internal.ServerImpl;
+import io.grpc.internal.ServerListener;
+import io.grpc.internal.ServerTransport;
+import io.grpc.internal.ServerTransportListener;
+import io.grpc.internal.SharedResourceHolder;
+import java.io.File;
+import java.io.IOException;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.ScheduledExecutorService;
+import javax.annotation.concurrent.NotThreadSafe;
+
+/**
+ * Builder to build a gRPC server that can run as a servlet.
+ */
+@NotThreadSafe
+public final class ServerBuilder extends AbstractServerImplBuilder<ServerBuilder> {
+
+  private ScheduledExecutorService scheduledExecutorService;
+  private boolean internalCaller;
+  private boolean usingCustomScheduler;
+
+  /**
+   * Builds a gRPC server that can run as a servlet.
+   *
+   * <p>The returned server will not been started or be bound a port.
+   *
+   * <p>Users should not call this method directly. Instead users should call {@link
+   * ServletAdapter.Factory#create(ServerBuilder)}, which internally will call {@code build()} and
+   * {@code start()} appropriately.
+   *
+   * @throws IllegalStateException if this method is called by users directly
+   */
+  @Override
+  public Server build() {
+    checkState(internalCaller, "method should not be called by the user");
+    return super.build();
+  }
+
+  ServerTransportListener buildAndStart() {
+    ServerImpl server;
+    try {
+      internalCaller = true;
+      server = (ServerImpl) build().start();
+      internalCaller = false;
+    } catch (IOException e) {
+      // actually this should never happen
+      throw new RuntimeException(e);
+    }
+
+    if (!usingCustomScheduler) {
+      scheduledExecutorService = SharedResourceHolder.get(GrpcUtil.TIMER_SERVICE);
+    }
+
+    // Create only one "transport" for all requests because it has no knowledge of which request is
+    // associated with which client socket. This "transport" does not do socket connection, the
+    // container does.
+    ServerTransportImpl serverTransport =
+        new ServerTransportImpl(scheduledExecutorService, usingCustomScheduler);
+    return server.getServerListener().transportCreated(serverTransport);
```

Comment:
I don't think we should have a `getServerListener()`. Is the only problem that `InternalServerImpl` can't call the listener during start?

Reply:
The only problem is to get access the `ServerListener` that the `ServerImpl` instance starts with, so that I can  reuse the `ServerImpl`'s code.

Reply:
So let's just have `InternalServerImpl` create the `ServerTransportImpl`.

Reply:
I populated the `serverListener` in `InternalServerImpl` back to the builder and got rid of it.